eras/conway/impl/src/Cardano/Ledger/Conway/Governance/Proposals.hs

```diff
@@ -1,132 +1,498 @@
+
+pnChildrenL :: Lens' (PNode a) (Set a)
+pnChildrenL = lens pnChildren $ \x y -> x {pnChildren = y}
+
+-- | A single proposal-tree. This map represents all the action-ids that
+-- form the tree.
+newtype PHierarchy a = PHierarchy
+  { unPHierarchy :: Map a (PNode a)
+  }
+  deriving stock (Show, Eq, Ord, Generic)
+  deriving newtype (NoThunks, NFData, Default)
+
+pHierarchyNodesL :: Lens' (PHierarchy a) (Map a (PNode a))
+pHierarchyNodesL = lens unPHierarchy $ \x y -> x {unPHierarchy = y}
+
+-- | A smart data-type that encapsulates the essence of having 4
+-- instance fields of the same parameterized data-structure. This is
+-- used to make the 4 proposals-forest roots, the 4 proposals-forest
+-- hierarchies and 4 `PrevGovActionIds`.
+data PForest (f :: Type -> Type) era = PForest
+  { pfPParamUpdate :: !(f (GovPurposeId 'PParamUpdatePurpose era))
+  , pfHardFork :: !(f (GovPurposeId 'HardForkPurpose era))
+  , pfCommittee :: !(f (GovPurposeId 'CommitteePurpose era))
+  , pfConstitution :: !(f (GovPurposeId 'ConstitutionPurpose era))
+  }
+
+deriving instance
+  ( Show (f (GovPurposeId 'PParamUpdatePurpose era))
+  , Show (f (GovPurposeId 'HardForkPurpose era))
+  , Show (f (GovPurposeId 'CommitteePurpose era))
+  , Show (f (GovPurposeId 'ConstitutionPurpose era))
+  ) =>
+  Show (PForest f era)
+
+pfPParamUpdateL :: Lens' (PForest f era) (f (GovPurposeId 'PParamUpdatePurpose era))
+pfPParamUpdateL = lens pfPParamUpdate $ \x y -> x {pfPParamUpdate = y}
+
+pfHardForkL :: Lens' (PForest f era) (f (GovPurposeId 'HardForkPurpose era))
+pfHardForkL = lens pfHardFork $ \x y -> x {pfHardFork = y}
+
+pfCommitteeL :: Lens' (PForest f era) (f (GovPurposeId 'CommitteePurpose era))
+pfCommitteeL = lens pfCommittee $ \x y -> x {pfCommittee = y}
+
+pfConstitutionL :: Lens' (PForest f era) (f (GovPurposeId 'ConstitutionPurpose era))
+pfConstitutionL = lens pfConstitution $ \x y -> x {pfConstitution = y}
+
+pfrPParamUpdateL :: Lens' (PForest PRoot era) (PRoot (GovPurposeId 'PParamUpdatePurpose era))
+pfrPParamUpdateL = pfPParamUpdateL
+
+pfrHardForkL :: Lens' (PForest PRoot era) (PRoot (GovPurposeId 'HardForkPurpose era))
+pfrHardForkL = pfHardForkL
+
+pfrCommitteeL :: Lens' (PForest PRoot era) (PRoot (GovPurposeId 'CommitteePurpose era))
+pfrCommitteeL = pfCommitteeL
+
+pfrConstitutionL :: Lens' (PForest PRoot era) (PRoot (GovPurposeId 'ConstitutionPurpose era))
+pfrConstitutionL = pfConstitutionL
+
+pfhPParamUpdateL :: Lens' (PForest PHierarchy era) (PHierarchy (GovPurposeId 'PParamUpdatePurpose era))
+pfhPParamUpdateL = pfPParamUpdateL
+
+pfhHardForkL :: Lens' (PForest PHierarchy era) (PHierarchy (GovPurposeId 'HardForkPurpose era))
+pfhHardForkL = pfHardForkL
+
+pfhCommitteeL :: Lens' (PForest PHierarchy era) (PHierarchy (GovPurposeId 'CommitteePurpose era))
+pfhCommitteeL = pfCommitteeL
+
+pfhConstitutionL :: Lens' (PForest PHierarchy era) (PHierarchy (GovPurposeId 'ConstitutionPurpose era))
+pfhConstitutionL = pfConstitutionL
```

Comment:
Why are these lenses necessary when we already have a more general lens above?

Reply:
Removed! :raised_hands: